cmd/snapshooter/main.go

```diff
@@ -0,0 +1,169 @@
+package main
+
+import (
+	"context"
+	"strconv"
+	"time"
+
+	elastic "github.com/olivere/elastic/v7"  // Elasticsearch client
+	"go.uber.org/zap"                        // Logging
+	kingpin "gopkg.in/alecthomas/kingpin.v2" // Command line args parser
+
+	"github.com/mintel/elasticsearch-asg/cmd" // Common logging setup func
+)
+
+const (
+	// Initial Elasticsearch exponential backoff retry time.
+	esRetryInit = 150 * time.Millisecond
+	// Max Elasticsearch exponential backoff retry time.
+	esRetryMax = 1200 * time.Millisecond
+)
+
+// SnapshotFormat is the format for snapshot names (time.Time.Format()).
+// Elasticsearch snapshot names may not contain spaces.
+const SnapshotFormat = "2006-01-02-15-04-05"
+
+// defaultURL is the default Elasticsearch URL.
+const defaultURL = "http://localhost:9200"
+
+// Command line opts
+var (
+	esURL   = kingpin.Arg("url", "Elasticsearch URL. Default: "+defaultURL).Default(defaultURL).URL()
+	windows = kingpin.Flag("window", "Snapshot frequency + TTL. May be set multiple times. ISO 8601 Duration string format. Example: `--window P1M=PT1H` == keep hourly snapshots for 1 month.").Default(
+		"P1M=PT1H",
+		"P3M=P1W",
+		"P3Y=P1M",
+	).StringMap()
+	delete       = kingpin.Flag("delete", "If set, clean up old snapshots. This is false by default for safety's sake.").Short('d').Bool()
+	repoName     = kingpin.Flag("repo", "Name of the snapshot repository.").Default("backups").String()
+	repoType     = kingpin.Flag("type", "If set, create a repository of this type before creating snapshots. See also: '--settings'").String()
+	repoSettings = kingpin.Flag("settings", "Use these settings creating the snapshot repository. May be set multiple times. Example: `--type=s3 --settings bucket=my_bucket`").StringMap()
+)
+
+func main() {
```

Comment:
Two instances of snapshooter with the same configuration shouldn't be dangerous, since they will deterministically choose when to create and delete snapshots. There would be a race condition between the instances, but the resulting snapshots [should be the same](https://sequencediagram.org/index.html#initialData=MoOwhgDgzgFg9nALgUwE4AICCBaAfAUQBswpEBLAYymTFQpgC4BhVGldKcaeRdAJgAMARgCc2AQHZsfAGzYhQ8QKUAoUJFgIUGAEJ4iJclRp1GLNsg5dNvQaPFTZ8xQOWuVB0pWq16eddxaaOg6DPioqHCoDOgBNugA7mSIMOgpYLzgALaWYISsYAAmAJ7oyAAeZKRQKiBIlqhkAOYwvHAAZrHW8PW6YZVtIGURUR7EXsa+MP7dQRiYDADyANIMcTzo7WQgVTDIhSpAA).

In any case, we [won't run more than one](https://gitlab.com/mintel/everest/terraform/blob/elasticsearch/modules/elasticsearch/snapshooter.tf#L83).

A simple cron wouldn't allow for complex logic like downsampling snapshots over time, but if we don't need complex logic (see my response to your other comment) then I agree it would be better.

Reply:
What's the downsampling snapshots? The way I'm thinking is the snapshooter could provide operation logic for snapshot creation and removal, but leave the schedule bit to cron.

Reply:
The downsampling is the removal logic, keeping more recent snapshots and fewer older snapshots ([example](https://github.com/mintel/elasticsearch-asg/blob/b4c1bf19442192c8f024efa5540044f59eca4ad9/cmd/snapshooter/README.md#example)).

Correct me if I'm wrong, but I think you're saying:
1. Have a cron that triggers snapshooter.
2. Snapshooter creates a snapshot at the time it is triggered.
3. After the snapshot is finished, snapshooter selects older snapshots to remove and deletes them.

If snapshooter does logic about which snapshots to remove, then I'm not in favour of cron because:
- We would have knowledge of snapshot scheduling in two places (the cron schedule for creating snapshots and snapshooter's schedule for removing them).
- The logic for which snapshots to remove would become more complex because snapshooter does not know the cron's schedule for creating snapshots.
- It's adds a [lot of Terraform](https://github.com/dxw/terraform-aws-ecs-scheduled-task/blob/master/main.tf).
- We haven't used scheduled ECS tasks anywhere else.

However, if we don't automatically remove snapshots, then I'm in favour of cron and snapshooter could be replaced by something as simple as `curl -X PUT "http://elasticsearch:9200/_snapshot/backups/$( date +"%y-%m-%d-%H-%M-%S" )?wait_for_completion=true"`.

Reply:
The way I'm thinking it could work is like:

As the equivalent of "P1M=PT1H" in the example

For example, an hourly cronjob to run, with an argument like
./snapshooter --keep-for 1M --period hourly

Which will trigger snapshooter to run the snapshot, then parse through the hourly backup to remove old snapshots.

It essentially is the same logic of one run of the for loop in your current code. The snapshooter code will not handle any scheduling, only take the information through argument.

The reason I'm thinking about this is the scheduling bit in current code make it impractical to run multiple instances (with different config) against the same ES. And there's no way to prevent that except you have to set up the deployment to make sure there's a single instance running.

The scheduling bit feels belong to a global context, instead of a utility-like container run on the side. An alternative direction which will make me feel better about keeping current snapshooter is to have a more centralized deployment setup for the whole system with mechanics to make sure the singularity of the snapshooter.

It's not a blocker issue for now, as I think the scheduler implemented here would work, but I feel it's a concern on the system design level